File: Plugin/AutoTrackPlugin/src/main/java/com/wallstreetcn/autotrack/DataScanTransform.java
package com.wallstreetcn.autotrack;

import com.android.build.api.transform.QualifiedContent;
import com.android.build.api.transform.Transform;
import com.android.build.api.transform.TransformException;
import com.android.build.api.transform.TransformInvocation;
import com.android.build.gradle.internal.pipeline.TransformManager;
import com.kronos.plugin.base.BaseTransform;
import com.kronos.plugin.base.ClassUtils;
import com.kronos.plugin.base.TransformCallBack;
import com.wallstreetcn.autotrack.scan.DataScanAsmHelper;

import org.jetbrains.annotations.NotNull;

import java.io.IOException;
import java.util.Set;

/**
 * @Author LiABao
 * @Since 2021/1/9
 */
public class DataScanTransform extends Transform {


    @Override
    public String getName() {
        return "DataScanTransform";
    }

    @Override
    public Set<QualifiedContent.ContentType> getInputTypes() {
        return TransformManager.CONTENT_JARS;
    }

    @Override
    public Set<? super QualifiedContent.Scope> getScopes() {
        return TransformManager.SCOPE_FULL_PROJECT;
    }

    @Override
    public boolean isIncremental() {
        return true;
    }

    @Override
    public void transform(TransformInvocation transformInvocation) throws TransformException, InterruptedException, IOException {
        final DataScanAsmHelper injectHelper = new DataScanAsmHelper();
        BaseTransform baseTransform = new BaseTransform(transformInvocation, new TransformCallBack() {

            @Override
            public byte[] process(@NotNull String className, byte[] bytes) {
                if (ClassUtils.checkClassName(className)) {
                    try {
                        return injectHelper.modifyClass(bytes);
                    } catch (Exception e) {
                        e.printStackTrace();
                    }
                }
                return null;
            }
        }, false);
        baseTransform.startTransform();
    }


}

File: Plugin/AutoTrackPlugin/src/main/java/com/wallstreetcn/autotrack/DataScanTransform.kt
package com.wallstreetcn.autotrack

import com.android.build.api.transform.QualifiedContent
import com.android.build.api.transform.Transform
import com.android.build.api.transform.TransformException
import com.android.build.api.transform.TransformInvocation
import com.android.build.gradle.internal.pipeline.TransformManager
import com.kronos.plugin.base.BaseTransform
import com.kronos.plugin.base.ClassUtils
import com.kronos.plugin.base.TransformCallBack
import com.wallstreetcn.autotrack.scan.DataScanAsmHelper
import java.io.IOException

/**
 * @Author LiABao
 * @Since 2021/1/9
 */
class DataScanTransform : Transform() {
    override fun getName(): String {
        return "DataScanTransform"
    }

    override fun getInputTypes(): Set<QualifiedContent.ContentType> {
        return TransformManager.CONTENT_JARS
    }

    override fun getScopes(): MutableSet<in QualifiedContent.Scope>? {
        return TransformManager.SCOPE_FULL_PROJECT
    }

    override fun isIncremental(): Boolean {
        return true
    }

    @Throws(TransformException::class, InterruptedException::class, IOException::class)
    override fun transform(transformInvocation: TransformInvocation) {
        val injectHelper = DataScanAsmHelper()
        val baseTransform = BaseTransform(transformInvocation, object : TransformCallBack {
            override fun process(className: String, bytes: ByteArray?): ByteArray? {
                if (ClassUtils.checkClassName(className)) {
                    try {
                        return injectHelper.modifyClass(bytes)
                    } catch (e: Exception) {
                        e.printStackTrace()
                    }
                }
                return null
            }
        })
        baseTransform.startTransform()
    }

}